packages/expo-image-picker/src/ExponentImagePicker.web.ts

```diff
@@ -139,68 +141,112 @@ function openFileBrowserAsync({
-          const image = new Image();
-          image.src = uri;
-          image.onload = () => {
-            returnMediaData({
-              uri,
-              width: image.naturalWidth ?? image.width,
-              height: image.naturalHeight ?? image.height,
-              type: 'image',
-              mimeType: targetFile.type,
-              fileName: targetFile.name,
-              fileSize: targetFile.size,
-            });
-          };
-          image.onerror = () => returnRaw();
-        } else if (targetFile.type.startsWith('video/')) {
-          const video = document.createElement('video');
-          video.preload = 'metadata';
-          video.src = uri;
-          video.onloadedmetadata = () => {
-            returnMediaData({
-              uri,
-              width: video.videoWidth,
-              height: video.videoHeight,
-              type: 'video',
-              mimeType: targetFile.type,
-              fileName: targetFile.name,
-              fileSize: targetFile.size,
-              duration: video.duration,
-            });
-          };
-          video.onerror = () => returnRaw();
-        } else {
-          returnRaw();
-        }
-      } else {
-        returnRaw();
+    reader.onload = (event: ProgressEvent<FileReader>) => {
+      const result = event.target?.result;
+      if (typeof result !== 'string') {
+        reject(new Error('Failed to read file as base64'));
+        return;
       }
+      // Remove the data URL prefix to get just the base64 data
+      resolve(result.substring(result.indexOf(',') + 1));
     };
-
-    reader.readAsDataURL(targetFile);
+    reader.readAsDataURL(file);
   });
 }
 
+/**
+ * Reads a file and returns its data as an ImagePickerAsset.
+ * Handles both base64 and blob URL modes, and extracts metadata for images and videos.
+ */
+async function readFile(targetFile: File, options: { base64: boolean }): Promise<ImagePickerAsset> {
+  const mimeType = targetFile.type;
+  const baseUri = URL.createObjectURL(targetFile);
+
+  try {
+    let metadata: { width: number; height: number; duration?: number };
+    let base64: string | undefined;
+
+    if (mimeType.startsWith('image/')) {
+      metadata = await getImageMetadata(baseUri);
+    } else if (mimeType.startsWith('video/')) {
+      metadata = await getVideoMetadata(baseUri);
+    } else {
+      throw new Error(`Unsupported file type: ${mimeType}. Only images and videos are supported.`);
+    }
+
+    if (options.base64) {
+      base64 = await readFileAsBase64(targetFile);
+    }
+
+    return {
+      uri: baseUri,
+      width: metadata.width,
+      height: metadata.height,
+      type: mimeType.startsWith('image/') ? 'image' : 'video',
+      mimeType,
+      fileName: targetFile.name,
+      fileSize: targetFile.size,
+      file: targetFile,
+      ...(metadata.duration !== undefined && { duration: metadata.duration }),
+      ...(base64 && { base64 }),
+    };
+  } catch (error) {
+    throw error;
```

Comment:
Unsupported files should throw instead of returning “raw”. This is now aligned with iOS and Android.